pyccl/halos/halo_model_base.py

```diff
@@ -0,0 +1,285 @@
+
+    def _check_mass_def(self, mass_def) -> None:
+        """ Return False if the input mass definition agrees with
+        the definitions for which this mass function parametrization
+        works. True otherwise. This function gets called at the
+        start of the constructor call.
+
+        Args:
+            mass_def (:class:`~pyccl.halos.massdef.MassDef`):
+                a mass definition object.
+        """
+        classname = self.__class__.__name__
+        msg = f"{classname} is not defined for {mass_def.name} masses"
+
+        if self._check_mass_def_strict(mass_def):
+            # Passed mass is incompatible with model.
+
+            if self._mass_def_strict_always:
+                # Class has no universal model and mass is incompatible.
+                raise ValueError(
+                    f"{msg} and this requirement cannot be relaxed.")
+
+            if self.mass_def_strict:
+                # Strict mass_def check enabled and mass is incompatible.
+                raise ValueError(
+                    f"{msg}. To relax this check set `mass_def_strict=False`.")
+
+    def _get_logM_sigM(self, cosmo, M, a, *, return_dlns=False):
+        """Compute ``logM``, ``sigM``, and (optionally) ``dlns_dlogM``."""
+        cosmo.compute_sigma()  # initialize sigma(M) splines if needed
+        logM = np.log10(M)
+
+        # sigma(M)
+        status = 0
+        sigM, status = lib.sigM_vec(cosmo.cosmo, a, logM, len(logM), status)
+        check(status, cosmo=cosmo)
+        if not return_dlns:
+            return logM, sigM
+
+        # dlogsigma(M)/dlog10(M)
+        dlns_dlogM, status = lib.dlnsigM_dlogM_vec(cosmo.cosmo, a, logM,
+                                                   len(logM), status)
+        check(status, cosmo=cosmo)
+        return logM, sigM, dlns_dlogM
+
+
+class MassFunc(HMIngredients):
+    """ This class enables the calculation of halo mass functions.
+    We currently assume that all mass functions can be written as
+
+    .. math::
+        \\frac{dn}{d\\log_{10}M} = f(\\sigma_M)\\,\\frac{\\rho_M}{M}\\,
+        \\frac{d\\log \\sigma_M}{d\\log_{10} M}
+
+    where :math:`\\sigma_M^2` is the overdensity variance on spheres with a
+    radius given by the Lagrangian radius for mass M.
+
+    * Subclasses implementing analytical mass function parametrizations
+      can be created by overriding the ``_get_fsigma`` method.
+
+    * Subclasses may have particular implementations of
+      ``_check_mass_def_strict`` to ensure consistency of the halo mass
+      definition.
+
+    Args:
+        mass_def (:class:`~pyccl.halos.massdef.MassDef`):
+            a mass definition object that fixes
+            the mass definition used by this mass function
+            parametrization.
+        mass_def_strict (bool): if False, consistency of the mass
+            definition will be ignored.
+    """
+    _mass_def_strict_always = False
+
+    @abstractmethod
+    def _get_fsigma(self, cosmo, sigM, a, lnM):
```

Comment:
Yes, I considered that on the Emulator PR (hence built the __linked_abstractmethods__ you didn't like). Let me think about that more and I'll get back to you on this. There's been changes to the abc module recently which we might want to take advantage of.

Reply:
Let's keep it simple, please. This isn't rocket science.

Reply:
I always aim to keep things simple, I'm sure this is some kind of trauma from the Caching PR.

Reply:
Okay, this is fixed now.

So, overall, `_real`, `_fourier`, `_projected`, `_cumul2d` were all missing from the base class, and that wasn't good. I created decorators that allow us to include them, while enforcing that one of [`_real`, `_fourier`] is implemented.

**Implementation details**
----------------------------------
**Abstract methods:**
The main issue is that we want to implement some abstract methods, because these tell the developer what **must** be implemented. To that end, if there is a single method that must be implemented, there is the `@abstractmethod` decorator from `abc`.

**Relaxing the abstraction criteria:**
Now, if we want an either/or, like is the case you point out here with `MassFunc`, `HaloBias`, `Concentration`, as well as `HaloProfile` (`_real`/`_fourier`) this is too much to be left to its own devices. We should have a mechanism that ensures that something is implemented.

So I created an `@abstractlinkedmethod` decorator which adds all methods that have it to a registry, `__isabstractlinkedmethod__`. This is exactly the same as what `abc` does with its own `@abstractmethod`. The difference here is that if *only one* of the abstract linked methods are implemented, the registry is cleared, and the class is ok to instantiate. I hope the `abc` developers will include that in the package at some point, because it's useful.

**HaloProfile is very special:**
A problem that then appears is in `HaloProfile`: to check e.g. whether the `_real` or the `_fourier_wrap` will be dispatched, there were attribute checks in place. And of course there would be, because they are inherited from the base class. So the checks shouldn't happen with attribute checks, but with abstraction checks (just like what the `inspect` module provides).

So there are now private methods that check if a method is abstract-linked or not, exactly mimicking what `inspect` does: check the hook. And the checks in `HaloProfile` change from `hasattr(self, "_real")` to `self._is_implemented("_real")`.

Finally, I noticed that there is no mention anywhere of the fact that developers may add their own `_projected` and `_cumul2d` methods. Just as `_real` and `_fourier`, these also have to be included as template methods, to show developers that they may be implemented (and that they are in fact used!).

To do that, I simply created another decorator `@templatemethod` that is used for templates. They have to have a decorator because this is how the hook is added and piped into `self._is_implemented`. This is how the profile understands that the method is actually implemented.

**Summary:**
All of this would be way easier if we weren't setting `_fourier_analytic` as `_fourier` etc. during instance creation in the matter profiles. We could just peek at the class and see whether it is abstract or not. But such is life.

Reply:
This is killing one bird with a hail of stones. We do not need this level of complexity for something as simple as this. We do not want to require all mass functions to have an f(sigma) parametrisation. That's all. In cases where this functional form does not exist, developers just overload the get_mass_function method.

Reply:
resolve if ok now

Reply:
Ok, I like this a lot more!